unified-runtime/source/loader/layers/sanitizer/asan/asan_shadow.cpp

```diff
@@ -282,51 +289,85 @@ ur_result_t ShadowMemoryGPU::AllocPrivateShadow(ur_queue_handle_t Queue,
                                                 uint64_t NumWI, uint32_t NumWG,
                                                 uptr *&Base, uptr &Begin,
                                                 uptr &End) {
-  {
-    const size_t Size = NumWI * sizeof(uptr);
-    ur_usm_desc_t Properties{UR_STRUCTURE_TYPE_USM_DESC, nullptr,
-                             UR_USM_ADVICE_FLAG_DEFAULT, sizeof(uptr)};
-    UR_CALL(getContext()->urDdiTable.USM.pfnDeviceAlloc(
-        Context, Device, &Properties, nullptr, Size, (void **)&Base));
-  }
+  // Trying to allocate private base array and private shadow, and any one of
+  // them fail to allocate would be a failure
+  ur_result_t UrRes;
+  static size_t LastPrivateBaseAllocedSize = 0;
+  static size_t LastPrivateShadowAllocedSize = 0;
+
+  do {
+    const size_t NewPrivateBaseSize = NumWI * sizeof(uptr);
+    if (NewPrivateBaseSize > LastPrivateBaseAllocedSize) {
+      if (PrivateBasePtr) {
+        UR_CALL_BREAK(UrRes, getContext()->urDdiTable.USM.pfnFree(
+                                 Context, (void *)PrivateBasePtr));
+        PrivateBasePtr = 0;
+        LastPrivateBaseAllocedSize = 0;
+      }
 
-  {
-    const size_t RequiredShadowSize =
+      ur_usm_desc_t PrivateBaseProps{UR_STRUCTURE_TYPE_USM_DESC, nullptr,
+                                     UR_USM_ADVICE_FLAG_DEFAULT, sizeof(uptr)};
+      UR_CALL_BREAK(UrRes, getContext()->urDdiTable.USM.pfnDeviceAlloc(
+                               Context, Device, &PrivateBaseProps, nullptr,
+                               NewPrivateBaseSize, (void **)&PrivateBasePtr));
+
+      // No need to clean the shadow base, their should be set by work item on
+      // launch
+
+      // FIXME: we should add private base to statistic
+      LastPrivateBaseAllocedSize = NewPrivateBaseSize;
```

Comment:
Good point, I think it can be counted as ShadowMalloced, no need to be so strict